Services/Tree/classes/Setup/class.ilTreeImplementationSwitch.php

```diff
@@ -0,0 +1,141 @@
+    {
+        return hash('sha256', self::class);
+    }
+
+    public function getLabel() : string
+    {
+        if ($this->mode === self::MATRIALIZED_PATH_TO_NESTED_SET) {
+            return 'The tree implementation is switched to nested set';
+        }
+
+        return 'The tree implementation is switched to materialized path';
+    }
+
+    public function isNotable() : bool
+    {
+        return true;
+    }
+
+    public function getPreconditions(Environment $environment) : array
+    {
+        return [
+            new ilSettingsFactoryExistsObjective(),
+            new ilDatabaseInitializedObjective(),
+        ];
+    }
+
+    public function achieve(Environment $environment) : Environment
+    {
+        $db = $environment->getResource(Environment::RESOURCE_DATABASE);
+        $settings_factory = $environment->getResource(Environment::RESOURCE_SETTINGS_FACTORY);
+
+        $is_nested_set = $settings_factory->settingsFor('common')->get('main_tree_impl', 'ns') === 'ns';
+
+        if ($this->mode === self::NESTED_SET_TO_MATERIALIZED_PATH && $is_nested_set) {
+            ilMaterializedPathTree::createFromParentReleation($db);
+
+            $db->dropIndexByFields('tree', ['lft']);
+            $db->dropIndexByFields('tree', ['path']);
+            $db->addIndex('tree', ['path'], 'i4');
+
+            $settings_factory->settingsFor('common')->set('main_tree_impl', 'mp');
+        } elseif ($this->mode === self::MATRIALIZED_PATH_TO_NESTED_SET && !$is_nested_set) {
+            $renumber_callable = function (ilDBInterface $db) {
+                $this->renumberNestedSet($db, 1, 1);
+            };
+            $ilAtomQuery = $db->buildAtomQuery();
+            $ilAtomQuery->addTableLock('tree');
+
+            $ilAtomQuery->addQueryCallable($renumber_callable);
+            $ilAtomQuery->run();
+
+            $db->dropIndexByFields('tree', ['lft']);
+            $db->dropIndexByFields('tree', ['path']);
+            $db->addIndex('tree', ['lft'], 'i4');
+
+            $settings_factory->settingsFor('common')->set('main_tree_impl', 'ns');
+        } else {
+            throw new UnachievableException('The tree implementation switch does already equal the requested mode');
+        }
+
+        return $environment;
+    }
+    
+    private function renumberNestedSet(ilDBInterface $db, int $node_id, int $i) : int
+    {
+        $db->manipulateF(
+            'UPDATE tree SET lft = %s WHERE child = %s',
+            ['integer', 'integer'],
+            [$node_id, $i]
+        );
+
+        $query = 'SELECT child FROM tree WHERE parent = ' . $db->quote($node_id, 'integer') . ' ORDER BY lft';
+        $res = $db->query($query);
+
+        $childs = [];
+        while ($row = $res->fetchRow(ilDBConstants::FETCHMODE_OBJECT)) {
+            $childs[] = (int) $row->child;
+        }
+
+        foreach ($childs as $child) {
+            $i = $this->renumberNestedSet($db, $child, $i + 1);
+        }
+        $i++;
+
```

Comment:
@mjansenDatabay the rgt value update is missing here.

Reply:
@smeyer-ilias DONE